kernel: provide Platform abstract interface for device drivers which are designed to works seamlessly on PCI bus, ACPI and DTS

najumon1980 opened the pull request:
Provide Platform abstract interface for device drivers which are designed to work on multiple bus topologies such as PCI, ACPI or just DTS base. Driver developers can use the new PLATFORM_DT_DEV_INST_DEFINE macro instead of DEVICE_DT_INST_DEFINE in case they have to design drivers which need to works seamlessly on PCI bus, ACPI and DTS base.  Currently  there is no support for multi MSI and multi vectors. Following are some of the key features/highlight:

- Abstract underlying bus/parent complexity and providing a generic interface to driver developers. Using the new interface, driver developers need not want to bother about underlying bus interface and associated complexity. 
- Based on their parent node type, the interface will automatically enumerate resource such as MMIO and IRQ resources and configure them which will be completely transparent to device drivers
- The interface will take care of memory map of MMIO address in case platform support MMU
- The new PLATFORM_DT_DEV_INST_DEFINE  interface take ISR as one of the argument and will take care the ISR registration based on underlying bus/parent node  type(PCI/ACPI or just DTS). 
- The new interface add very little overhead (almost NIL)

najumon1980 committed:
kernel: devices: add platform abstract support

add abstract interface for devices drivers to enumerate and
configure resource such as mmio and interrupt. This will help
drivers which are designed to work on various bus such as pci,
acpi or can me dts base.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
board: x86: rpl: add acpi resource enumeration for uart driver

add acpi resource enumeration for ns16550 uart driver on Intel
raptor lake platform

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
drivers: uart: ns16550: update with new platform interface

update ns16550 uart driver with new platform interface for
enumerate and configure device resources. This help the drivers
works on multiple bus topology such as pci, acpi or dts base.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
boards: ns16550: update dts with io mapped flags

update board dts files with io_mapped flag where ns16550 register
base is io mapped devices

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
drivers: rtc: add acpi based resource enumeration

add acpi based resource enumeration for retrieve rtc mmio and
interrupt resources

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
arch: x86: disable early console if uart use acpi

disable early console if uart use acpi based resource enumeration

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
lib: acpi: add header for dts bindings

add acpi header for dts bindings for various acpi dts defintions

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 committed:
dts: base: add io-mapped flag for iomapped devices

add dts flag for distinguish iomapped devices vs memory mapped device

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

dcpleung reviewed (changes requested):
I still think this is complicating things too much. Can this be flowed into `DEVICE_MMIO_*` so you are not introducing yet another API?

najumon1980 committed:
boards: intel: update dts with io mapped flags and acpi hid for rtc

update board dts files with io_mapped flag where rtc register
base is io mapped devices. Also update acpi hid for ehl, rpl and
adl platforms.

Signed-off-by: Najumon B.A <najumon.ba@intel.com>

najumon1980 commented on dts/bindings/base/base.yaml:
```diff
@@ -23,6 +23,10 @@ properties:
+  io-mapped:
+    type: boolean
+    description: specify registers are IO mapped or memory mapped
+
```
moved to a new yaml file

najumon1980 commented on include/zephyr/platform.h:
```diff
@@ -0,0 +1,309 @@
+/*
```
updated

najumon1980 commented on include/zephyr/platform.h:
```diff
@@ -0,0 +1,309 @@
+ */
+#define PLATFORM_MMIO_RAM                                                                          \
+	DEVICE_MMIO_RAM;                                                                           \
+	struct pcie_dev *pcie
```
macro defined based on node position (parent node) and pcie struct is only if it is on PCIe bus/parent node (and if not it simply defined to DEVICE_MMIO_RAM)

najumon1980 commented on include/zephyr/platform.h:
```diff
@@ -0,0 +1,309 @@
+#if DT_ANY_INST_ON_BUS_STATUS_OKAY(pcie)
+#define PLATFORM_PCI_DECLARE_0(n)
+#define PLATFORM_PCI_DECLARE_1(n) DEVICE_PCIE_INST_DECLARE(n);
+#define Z_PLATFORM_PCI_DECLARE(n) _CONCAT(PLATFORM_PCI_DECLARE_, DT_INST_ON_BUS(n, pcie))(n)
```
updated

najumon1980 commented:
> I still think this is complicating things too much. Can this be flowed into `DEVICE_MMIO_*` so you are not introducing yet another API?

my suggestion is better to be a new API and drivers which need to enable on PCI/ACPI can make use of this new API.

dcpleung commented:
> > I still think this is complicating things too much. Can this be flowed into `DEVICE_MMIO_*` so you are not introducing yet another API?
> 
> my suggestion is better to be a new API and drivers which need to enable on PCI/ACPI can make use of this new API.

But you are introducing another API with prefix `PLATFORM_` which absolutely adds to the confusion. Devices are devices, no matter which buses they are on.

najumon1980 commented:
> > > I still think this is complicating things too much. Can this be flowed into `DEVICE_MMIO_*` so you are not introducing yet another API?
> > 
> > 
> > my suggestion is better to be a new API and drivers which need to enable on PCI/ACPI can make use of this new API.
> 
> But you are introducing another API with prefix `PLATFORM_` which absolutely adds to the confusion. Devices are devices, no matter which buses they are on.

Agree and we can use same name (or flowed into) as standard DEVICE_XX macros. But my intention was that we need to distinguish these macros from standard DEVICE_XX macros and driver which plan to use in PCI bus or ACPI apart from DTS base shall only use these macros. Also I just renamed these macros to BUS_DEV_XX which avoid/reduce confusion

najumon1980 commented:
@dcpleung updated as per your comments and also responded to your comments as well.

dcpleung commented:
Changing the prefix from `PLATFORM_` to `BUS_` doesn't really solve the underlying issue: that you are introducing a new, separate/parallel way for device initialization. Both the NS16550 and RTC drivers (in this PR) require changes to utilize this. 

Do you mind sending a RFC to the mailing list to gather more comments on this?
